tools/gfx-unit-test/instanced-draw-tests.cpp

```diff
@@ -0,0 +1,581 @@
+            return indexBuffer;
+        }
+
+        void setUpAndDraw()
+        {
+            createRequiredResources();
+
+            ICommandQueue::Desc queueDesc = { ICommandQueue::QueueType::Graphics };
+            auto queue = device->createCommandQueue(queueDesc);
+            auto commandBuffer = transientHeap->createCommandBuffer();
+
+            auto encoder = commandBuffer->encodeRenderCommands(renderPass, framebuffer);
+            auto rootObject = encoder->bindPipeline(pipelineState);
+
+            gfx::Viewport viewport = {};
+            viewport.maxZ = 1.0f;
+            viewport.extentX = kWidth;
+            viewport.extentY = kHeight;
+            encoder->setViewportAndScissor(viewport);
+
+            encoder->setVertexBuffer(0, vertexBuffer, sizeof(Vertex));
+            encoder->setVertexBuffer(1, instanceBuffer, sizeof(Instance));
+            encoder->setIndexBuffer(indexBuffer, Format::R32_UINT);
+            encoder->setPrimitiveTopology(PrimitiveTopology::TriangleList);
+
+            uint32_t maxDrawCount = 1;
+            uint64_t argOffset = offsetof(IndexedIndirectArgData, args);
+            uint64_t countOffset = offsetof(IndexedIndirectArgData, count);
+
+            encoder->drawIndexedIndirect(maxDrawCount, indirectBuffer, argOffset, indirectBuffer, countOffset);
+            encoder->endEncoding();
+            commandBuffer->close();
+            queue->executeCommandBuffer(commandBuffer);
+            queue->waitOnHost();
+        }
+
+        void run()
+        {
+            indexBuffer = createIndexBuffer(device);
+            indirectBuffer = createIndirectBuffer(device);
+
+            setUpAndDraw();
+
+            const int kPixelCount = 4;
+            const int kChannelCount = 4;
+            int testXCoords[kPixelCount] = { 64, 192, 64, 192 };
+            int testYCoords[kPixelCount] = { 32, 100, 150, 250 };
+            float testResults[kPixelCount * kChannelCount];
+
+            checkTestResults(kPixelCount, kChannelCount, testXCoords, testYCoords, testResults);
+        }
+    };
+
+    void drawInstancedTestImpl(IDevice* device, UnitTestContext* context)
+    {
+        DrawInstancedTest test;
+        test.init(device, context);
+        test.run();
+    }
+
+    void drawIndexedInstancedTestImpl(IDevice* device, UnitTestContext* context)
+    {
+        DrawIndexedInstancedTest test;
+        test.init(device, context);
+        test.run();
+    }
+
+    void drawIndirectTestImpl(IDevice* device, UnitTestContext* context)
+    {
+        DrawIndirectTest test;
+        test.init(device, context);
+        test.run();
+    }
+
+    void drawIndexedIndirectTestImpl(IDevice* device, UnitTestContext* context)
+    {
+        DrawIndexedIndirectTest test;
+        test.init(device, context);
+        test.run();
+    }
+
+    SLANG_UNIT_TEST(drawInstancedD3D12)
+    {
+        runTestImpl(drawInstancedTestImpl, unitTestContext, Slang::RenderApiFlag::D3D12);
```

Comment:
Along those lines, I think I'll at least add a templated `drawTestImpl()` for `drawInstancedTestImpl`, `drawIndexedInstancedTestImpl`, etc. since all of them just create the appropriate test struct, call `init()`, then call `run()`.